torch/nn/modules/transformer.py

```diff
@@ -267,6 +267,24 @@ class TransformerEncoderLayer(Module):
     Neural Information Processing Systems, pages 6000-6010. Users may modify or implement
     in a different way during application.
 
+    forward() will use a special optimized implementation if all of the following
+    conditions are met:
+    - inference mode is enabled (using the torch.inference_mode() context manager)
+    - training is disabled (using .eval())
```

Comment:
Since this operation is supposed to be semantically equivalent when no gradients are run, I don't think it's necessary to test for eval(). The main reason for eval() is that we have some modules (like dropout) that literally do different things depending on if they're in training or eval mode. This does the same thing (but doesn't support backprop), so the reasoning for this doesn't apply.

Reply:
NestedTensor cannot be used in training because it doesn't support grad, and we don't support dropout in the native implementation, so the fast path definitely will not work in training; does this meet the bar for a .eval() check?

Reply:
OK